src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -862,12 +943,50 @@ static ucs_status_t uct_tcp_sockcm_ep_server_create(uct_tcp_sockcm_ep_t *tcp_ep,
               tcp_ep->fd, params_tcp_sockcm, tcp_ep->state);
 
     /* now that the server's ep was created, can try to send data */
-    ucs_async_modify_handler(tcp_ep->fd, UCS_EVENT_SET_EVWRITE | UCS_EVENT_SET_EVREAD);
-    return UCS_OK;
+    ucs_async_modify_handler(tcp_ep->fd, UCS_EVENT_SET_EVWRITE |
+                                         UCS_EVENT_SET_EVREAD);
+
```

Comment:
here you ask to be notified of when you can write to the fd (write event), but next, in uct_tcp_sockcm_ep_connect below, you call uct_tcp_sockcm_ep_progress_send to send even without waiting for this write event.
AFAIR we agreed that all sends and receives would be through the async thread events.

Reply:
seems like I missed that, I'll see how it works, thanks!

Reply:
fixed

Reply:
what if you get the write event on the async thread before doing the packing below? there will be events all the time until you do the actual packing.
maybe move lines 945-946 to the end of the function or move the entire section below to after getting the write event?

Reply:
this is under `UCS_ASYNC_BLOCK`, should not it solve the issue?

Reply:
yes it should, didnt notice it was under it.
uct_tcp_sockcm_ep_server_create becomes very long, can it be broken down?

Reply:
> uct_tcp_sockcm_ep_server_create becomes very long, can it be broken down?

see conversation https://github.com/openucx/ucx/pull/6503#discussion_r600450168